Validate URL hash

mattesCZ opened the pull request:
Resolves #5131 

Validates Latitude coordinate, that can trigger an error if out of [-90, 90]. Also added some checks for Zoom, Bearing and Pitch, that can be helpful, but they're not necessary.

## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->


 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [ ] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

HarelM commented on src/ui/hash.test.ts:
```diff
@@ -115,6 +115,21 @@ describe('hash', () => {
 
         expect(hash._onHashChange()).toBeFalsy();
+
+        // Zoom out of bounds, default maxZoom 22
```
Better to split this test to several tests and use these comments as the tests names.

mattesCZ replied:
Ok, I've just added some other invalid examples to the list. But maybe completely new test for `_isValidHash` is a way to go.

HarelM commented on src/ui/hash.ts:
```diff
@@ -103,7 +104,7 @@ export class Hash {
     _onHashChange = () => {
         const loc = this._getCurrentHash();
-        if (loc.length >= 3 && !loc.some(v => isNaN(v))) {
+        if (this._isValidHash(loc)) {
```
Can we have a better name than "loc"? I know this was not changed as part of this PR, but still...

HarelM replied:
Sure.

mattesCZ commented on src/ui/hash.ts:
```diff
@@ -150,4 +151,24 @@ export class Hash {
+
+    _isValidHash(loc: number[]) {
+        if (loc.length >= 3 && !loc.some(isNaN)) {
+            try {
```
I wanted to reuse code in `LngLat`, that throws error if latitude is out of range.

HarelM replied:
Then please wrap only the part where you expect the exception to be thrown.

HarelM commented on src/ui/hash.test.ts:
```diff
@@ -329,6 +325,44 @@ describe('hash', () => {
+    });
+
+    test('#_hasValidHash invalid', () => {
+        const hash = createHash()
```
Consider moving this setup to `beforeEach`.

mattesCZ replied:
It's possible, if I make nested `describe` block.

HarelM commented on src/ui/hash.ts:
```diff
@@ -102,14 +103,14 @@ export class Hash {
-        const loc = this._getCurrentHash();
-        if (loc.length >= 3 && !loc.some(v => isNaN(v))) {
-            const bearing = this._map.dragRotate.isEnabled() && this._map.touchZoomRotate.isEnabled() ? +(loc[3] || 0) : this._map.getBearing();
+        if (this._hasValidHash()) {
```
Consider inverting this as well.

HarelM commented on src/ui/hash.ts:
```diff
@@ -102,14 +103,14 @@ export class Hash {
-        if (loc.length >= 3 && !loc.some(v => isNaN(v))) {
-            const bearing = this._map.dragRotate.isEnabled() && this._map.touchZoomRotate.isEnabled() ? +(loc[3] || 0) : this._map.getBearing();
+        if (this._hasValidHash()) {
+            const hash = this._getCurrentHash();
```
I'm not sure if getCurrentHash method is expensive, but if it is I would consider calling it outside and pass the hash value, make the method a pure function.

HarelM commented:
I've added a few minor nit-picking. 
Looks good otherwise, thanks!!

mattesCZ commented on src/ui/hash.test.ts:
```diff
@@ -329,6 +325,44 @@ describe('hash', () => {
+        expect(hash._hasValidHash()).toBeTruthy();
+    });
+
+    test('#_hasValidHash invalid', () => {
```
I agree, I wanted to make it similar to the current code, not adding extra new concepts when I'm just making a bugfix.

mattesCZ commented:
> I've added a few minor nit-picking. Looks good otherwise, thanks!!

Thanks, I've refactored it a little, but I'm not sure about calling `hash._getCurrentHash()` in tests. Let me know if it's fine.

HarelM reviewed (approved):
LGTM, THANKS!
